docs/configurations.md
# Configuration management endpoints

Managing configurations is done at the "/configuration" endpoint, and
managing a *specific* configuration is done at the "/configuration/:name"
endpoint. You must be an authenticated user to access the data.

The following operations are available:

## GET /configuration

Retrieves a list of the available configurations.

You can add `sort=<property>` to sort the entries by that property.
I.e., adding `?sort=name` to the URI will sort the entries by their
name. Adding `dir=[asc|desc]` controls whether they are sorted into
ascending or descending order.

You can also add `page=<num>` to request a specific page of the data. By
default, it will give you 10 records at a time, but you can add
\`num=<num>` to the querystring to request a specific number of records.
When the results are paginated, there will be a special `links` key in
the returned results that will contain links to the next and previous
pages.

Example:

```bash
$ curl -H "Authorization: <token here>" https://<host>:<port>/configuration
{"configurations":[{"name":"host1","hostname":"nessus-ntp.lab.com","port":1241,"username":"toto"},{"name":"host2","hostname":"nessus-xml.lab.com","port":10000,"username":"admin"}]}
$ # with sorting
$ curl -H "Authorization: <token here>" https://<host>:<port>/configuration?sort=port&dir=desc
{"configurations":[{"name":"host2","hostname":"nessus-xml.lab.com","port":10000,"username":"admin"},{"name":"host1","hostname":"nessus-ntp.lab.com","port":1241,"username":"toto"}]}
```

## POST /configuration

Creates a new configuration.

Content of the new configuration object is in JSON format in the body of
the request. Will return 201 if the resource is successfully created,
and the URI for the resource will be returned in a JSON formatted
object.

Example:

```bash
$ curl -H "Authorization: <token here>" -X POST -d '{"name": "host3", "hostname": "foo.example.com", "port": 1234, "username": "bar"}' https://<host>:<port>/configuration
{"configuration": "/configuration/host3"}
```

## GET /configuration/:name

Retrieves a single configuration object.

Example:

```bash
$ curl -H "Authorization: <token here>" https://<host>:<port>/configuration/host1
{"name":"host1","hostname":"nessus-ntp.lab.com","port":1241,"username":"toto"}
```

## PUT /configuration/:name

Updates an existing resource.

The request body should be a new version of the resource found at the
endpoint, in JSON format. Will return a 404 if the existing resource is
not found, or 204 if the resource is successfully updated.

Example:

```bash
$ curl -H "Authorization: <token here>" -X PUT -d '{"name":"host1","hostname":"nessus-ntp.lab.com","port":1241,"username":"dorothy"}' https://<host>:<port>/configuration/host1

```

## DELETE /configuration/:name

Deletes an existing resource.

Will return 204 if the resource is successfully deleted.

Example:

```bash
$ curl -H "Authorization: <token here>" -X DELETE https://127.0.0.1/configuration/host1

```


config.js
module.exports = {
    URL: "/configuration",
    all: function(app) {
        var configs = app.configs
        return function(args) {
            var repsonse,
                params = args.params,
                sort = params["sort"] || "name",
                dir = params["dir"] || "asc",
                page = params["page"],
                num = params["num"],
                opts = {
                    "sort": sort,
                    "dir": dir,
                    "page": page,
                    "num": num
                };
            var all_configs = configs.read("*", opts, function(err, result) {
                response = app.json_response({ "configurations": result });
            });
            return response;
        };
    },
    create: function(app) {
        var configs = app.configs,
            self = this;
        return function(args) {
            var response,
                data = args.data;
            configs.read({ "name": data.name }, function(err, result) {
                if (result.length > 0) {
                    response = app.error(400, "Bad Request");
                } else {
                    configs.create(data);
                    var obj = JSON.stringify({ configuration: self.URL + "/" + data.name });
                    response = {
                        status: 201,
                        content: obj,
                        headers: {
                            "Content-Type": "application/json",
                            "Content-Length": obj.length
                        }
                    };
                }
            });
            return response;
        };
    },
    get: function(app) {
        var configs = app.configs
        return function(args) {
            var response,
                captures = args.captures,
                name = captures[0],
                config = configs.read({ "name": name }, function(err, result) {
                    if (err) {
                        throw err;
                    }

                    if (result.length < 1) {
                        response = app.error(404, "Not Found");
                    } else {
                        response = app.json_response(result[0]);
                    }
                });
            return response;
        };
    },
    update: function(app) {
        var configs = app.configs
        return function(args) {
            var response,
                captures = args.captures,
                name = captures[0],
                data = args.data;

            if (configs.update({ "name": name }, data)) {
                return {
                    status: 204,
                    content: ""
                };
            } else {
                return app.error(404, "Not Found");
            }
        };
    },
    delete: function(app) {
        var configs = app.configs
        return function(args) {
            var response,
                captures = args.captures,
                name = captures[0];
            var deleted = configs.delete({ "name": name });
            if (deleted) {
                return {
                    status: 204,
                    content: ""
                };
            } else {
                return app.error(404, "Not Found");
            }
        };
    }
};

